membership-attribute-service/app/services/ZuoraAttributeService.scala

```diff
@@ -0,0 +1,141 @@
+package services
+import com.gu.memsub.Subscription.AccountId
+import com.gu.memsub.subsv2.Subscription
+import com.gu.memsub.subsv2.SubscriptionPlan.AnyPlan
+import com.gu.memsub.subsv2.reads.SubPlanReads
+import com.gu.memsub.subsv2.reads.SubPlanReads.anyPlanReads
+import com.gu.zuora.ZuoraRestService.QueryResponse
+import loghandling.LoggingField.LogField
+import loghandling.{LoggingWithLogstashFields, ZuoraRequestCounter}
+import models.Attributes
+import org.joda.time.LocalDate
+import play.api.libs.concurrent.Execution.Implicits._
+
+import scala.concurrent.Future
+import scalaz.std.list._
+import scalaz.std.scalaFuture._
+import scalaz.syntax.traverse._
+import scalaz.{-\/, Disjunction, EitherT, \/, \/-, _}
+
+
+class ZuoraAttributeService(identityIdToAccountIds: String => Future[String \/ QueryResponse],
+                            subscriptionsForAccountId: AccountId => SubPlanReads[AnyPlan] => Future[Disjunction[String, List[Subscription[AnyPlan]]]],
+                            dynamoAttributeGetter: String => Future[Option[Attributes]]) extends LoggingWithLogstashFields {
+
+  def get(userId: String): Future[Option[Attributes]] = attributesFromZuora(userId)
+
+  def attributesFromZuora(identityId: String): Future[Option[Attributes]] = {
+
+    val attributesDisjunction = for {
+      accounts <- EitherT[Future, String, QueryResponse](withTimer(s"ZuoraAccountIdsFromIdentityId", zuoraAccountsQuery(identityId, identityIdToAccountIds), identityId))
+      accountIds = queryToAccountIds(accounts)
+      subscriptions <- EitherT[Future, String, List[Subscription[AnyPlan]]](
+        if(accountIds.nonEmpty) withTimer(s"ZuoraGetSubscriptions", getSubscriptions(accountIds, identityId), identityId)
+        else Future.successful(\/.right {
+          log.info(s"User with identityId $identityId has no accountIds and thus no subscriptions.")
+          Nil
+        })
+      )
+    } yield {
+      AttributesMaker.attributes(identityId, subscriptions, LocalDate.now())
```

Comment:
LocalDate.now() is another side effect that would need passing in before it's testable.  The amount of times we've had a test failing due to passing a date is quite high!

Reply:
Well spotted. I'm on it; thanks.